toolchain/lex/lex.cpp

```diff
@@ -1297,15 +1231,158 @@ auto Lexer::LexFileEnd(llvm::StringRef source_text, ssize_t position) -> void {
+    if (kind.is_opening_symbol()) {
+      open_groups_.push_back(token);
+      continue;
+    }
+
+    if (!kind.is_closing_symbol()) {
+      continue;
+    }
+
+    // Find the innermost matching opening symbol.
+    auto opening_it = std::find_if(
+        open_groups_.rbegin(), open_groups_.rend(),
+        [&](TokenIndex opening_token) {
+          return buffer_.GetTokenInfo(opening_token).kind.closing_symbol() ==
+                 kind;
+        });
+    if (opening_it == open_groups_.rend()) {
+      auto& token_info = buffer_.GetTokenInfo(token);
+      token_info.kind = TokenKind::Error;
+      token_info.error_length = kind.fixed_spelling().size();
+      fixed_anything = true;
+
+      CARBON_DIAGNOSTIC(
+          UnmatchedClosing, Error,
+          "Closing symbol without a corresponding opening symbol.");
+      token_emitter_.Emit(token, UnmatchedClosing);
+      continue;
+    }
+
+    // All intermediate open tokens have no matching close token.
+    for (auto it = open_groups_.rbegin(); it != opening_it; ++it) {
+      handle_missing_close(*it, token);
+    }
+    open_groups_.erase(opening_it.base() - 1, open_groups_.end());
+  }
+
+  // Diagnose any remaining unmatched opening symbols.
+  for (auto token : open_groups_) {
+    handle_missing_close(token, TokenIndex::Invalid);
+  }
+
+  // Merge the recovery tokens into the token list.
+  CARBON_CHECK(fixed_anything) << "Didn't find anything to fix";
+  auto old_tokens = std::move(buffer_.token_infos_);
+  buffer_.token_infos_.clear();
+  buffer_.token_infos_.reserve(old_tokens.size() + new_tokens.size());
+
+  int old_tokens_offset = 0;
+  for (auto [next_offset, info] : new_tokens) {
+    buffer_.token_infos_.append(old_tokens.begin() + old_tokens_offset,
+                                old_tokens.begin() + next_offset);
+    buffer_.token_infos_.push_back(info);
+    old_tokens_offset = next_offset;
+  }
+  buffer_.token_infos_.append(old_tokens.begin() + old_tokens_offset,
+                              old_tokens.end());
+
+  // Update the token list with closing and opening tokens. Note that this
+  // must be done last, and must be redone for all tokens, because token
+  // indexes were changed in the previous step.
+  open_groups_.clear();
+  for (auto token : buffer_.tokens()) {
+    auto kind = buffer_.GetKind(token);
+    if (kind.is_opening_symbol()) {
+      open_groups_.push_back(token);
+    } else if (kind.is_closing_symbol()) {
+      CARBON_CHECK(!open_groups_.empty()) << "Failed to balance brackets";
+      auto opening_token = open_groups_.pop_back_val();
+
+      CARBON_CHECK(kind ==
+                   buffer_.GetTokenInfo(opening_token).kind.closing_symbol())
+          << "Failed to balance brackets";
+      auto& opening_token_info = buffer_.GetTokenInfo(opening_token);
+      auto& closing_token_info = buffer_.GetTokenInfo(token);
+      opening_token_info.closing_token = token;
+      closing_token_info.opening_token = opening_token;
+    }
+  }
```

Comment:
Maybe extract this into a method to apply a list of token updates?

Reply:
Done.